src/Data/Text/Internal/Encoding/Utf8.hs

```diff
@@ -52,6 +53,21 @@ between :: Word8                -- ^ byte to check
 between x y z = x >= y && x <= z
 {-# INLINE between #-}
 
+-- TODO make branchless by looking into Word64 by clz (ord c)
+utf8Length :: Char -> Int
+utf8Length c
+  | ord c < 0x80    = 1
+  | ord c < 0x800   = 2
+  | ord c < 0x10000 = 3
+  | otherwise       = 4
```

Comment:
I know this has an optimised implementation now, but it looks more complex than something like 

```haskell
utf8Length (C# c) = I# (1# + geChar# c 0x80 + geChar c 0x800 + geChar# c 0x10000)
```

Just curious how this would compare (I might have the details wrong RE whether ti should be gtChar# or geChar#)

Reply:
@axman6-da thanks, that's an excellent suggestion! Indeed, `utf8Length (C# c) = I# (1# +# geChar# c (chr# 0x80#) +# geChar# c (chr# 0x800#) +# geChar# c (chr# 0x10000#))` works well and must be a bit faster. I'll incorporate this change later this week.

```
-- RHS size: {terms: 18, types: 3, coercions: 0, joins: 0/0}
utf8Length :: Char -> Int
utf8Length
  = \ (ds_d3gC :: Char) ->
      case ds_d3gC of { C# c_a2NH ->
      I#
        (+#
           1#
           (+#
              (geChar# c_a2NH '\65536'#)
              (+# (geChar# c_a2NH '\2048'#) (geChar# c_a2NH '\128'#))))
      }
```

Reply:
Done, thanks!